src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -279,31 +279,91 @@ private void updateOutOfSyncInvoice(Invoice invoice) {
+  }
+
+  private CompletableFuture<Void> updateWithSystemGeneratedData(Invoice invoice) {
+    return generateFolioInvoiceNumber()
+      .thenApply(invoice::withFolioInvoiceNo)
+      .thenAccept(this::calculateTotals)
+      .thenAccept(ok -> generateAdjustmentsIds(invoice));
+  }
+
+  private void generateAdjustmentsIds(Invoice invoice) {
+    invoice.getAdjustments()
+      .stream()
+      .filter(adj -> adj.getId() == null)
+      .forEach(adjustment -> adjustment.setId(randomUUID().toString()));
+  }
+
+  /**
+   * Updates total values of the invoice
+   * @param invoice invoice to update totals for
+   * @param lines invoice lines for the invoice
+   * @return {code true} if adjustments total, sub total or grand total value is different to original one
+   */
+  private boolean recalculateTotals(Invoice invoice, List<InvoiceLine> lines) {
+    // 1. Get original values
+    Double total = invoice.getTotal();
+    Double subTotal = invoice.getSubTotal();
+    Double adjustmentsTotal = invoice.getAdjustmentsTotal();
+
+    // 2. Recalculate totals
+    calculateTotals(invoice, lines);
+
+    // 3. Compare if anything has changed
+    return !(Objects.equals(total, invoice.getTotal()) && Objects.equals(subTotal, invoice.getSubTotal())
+        && Objects.equals(adjustmentsTotal, invoice.getAdjustmentsTotal()));
   }
 
   private CompletableFuture<Boolean> recalculateTotals(Invoice updatedInvoice, Invoice invoiceFromStorage) {
     // If invoice was approved, the totals are already fixed and should not be recalculated
     if (isPostApproval(invoiceFromStorage)) {
       return completedFuture(false);
     }
-    return recalculateTotals(updatedInvoice);
+
+    return processProratedAdjustments(updatedInvoice, invoiceFromStorage).thenCompose(ok -> recalculateTotals(updatedInvoice));
+  }
+
+  private CompletableFuture<Void> processProratedAdjustments(Invoice updatedInvoice, Invoice invoiceFromStorage) {
+    List<Adjustment> currentAdjustments = getProratedAdjustments(updatedInvoice);
+
+    // Skip if prorated adjustments are the same in incoming invoice and from storage
+    if (CollectionUtils.isEqualCollection(currentAdjustments, getProratedAdjustments(invoiceFromStorage))) {
+      return completedFuture(null);
+    }
+
+    return fetchInvoiceLinesByInvoiceId(updatedInvoice.getId()).thenCompose(lines -> {
+      if (lines.isEmpty()) {
+        return completedFuture(null);
+      }
+
+      // Collecting original hash codes of each invoice line record.
+      List<Integer> originalLineHashCodes = calculateHashCodes(lines);
+
+      invoiceLineHelper.processProratedAdjustments(lines, updatedInvoice);
+
+      return VertxCompletableFuture.allOf(ctx, lines.stream()
+        // If any update is made to invoice line adjustment(s), the new hash code will be generated for invoice line record
+        .filter(line -> !originalLineHashCodes.contains(line.hashCode()))
```

Comment:
There is no guarantee that after the line change the new hash code will not match the hash code of another line on the list. Since this is related to monetary calculations, I would suggest using a more reliable filtering method.

Reply:
Changes applied to use array list which holds invoice lines affected by prorated adjustment and may contain duplicates. When final result is returned, distinct values are only taken into account